### PR TITLE
cuBlasLt Epilogue To Fuse Linear + ReLU|GeLU

### PR DESCRIPTION
**Pull request by mingxu1067**
### PR types

New features

### PR changes

OPs

### Describe

1. Added `fused_gemm_epilogue_op` to compute Matmul+ ElementwiseAdd + ReLU|GeLU.
2. Added `fused_gemm_epilogue_grad_op` to compute ElementwiseAdd_grad + Matmul_grad+ [ReLU|GeLU]_grad.
3. Added a class member `fuse_gemm_epilogue` to `BuildStrategy` to `enable fuse_gemm_epilogue_pass`.
4. CUDA 11.6+ is required.
5. Usage Example
```python
class MultiFCLayer(paddle.nn.Layer):
    def __init__(self, hidden, Activation):
        super(MultiFCLayer, self).__init__()
        self.linear1 = paddle.nn.Linear(hidden, hidden)
        self.linear2 = paddle.nn.Linear(hidden, hidden)
        self.linear3 = paddle.nn.Linear(hidden, hidden)

        self.relu1 = Activation()
        self.relu2 = Activation()
        self.relu3 = Activation()

    def forward(self, x, matmul_y, ele_y):
        output = self.linear1(x)
        output = self.relu1(output)
        output = self.linear2(output)

        output1 = paddle.matmul(output, matmul_y)
        output = self.linear3(output)
        output = self.relu2(output)

        output = paddle.matmul(output, matmul_y)
        output = paddle.add(output, ele_y)
        output = self.relu3(output)
        output = paddle.add(output, output1)
        return output

paddle.enable_static()

batch = 64
seqlen = 128
hidden = 768

main_prog = paddle.static.Program()
startup_prog = paddle.static.Program()

with paddle.static.program_guard(main_prog, startup_prog):
    data = paddle.static.data(
        name="_data",
        shape=[-1, seqlen, hidden],
        dtype='float32')
    matmul_y = paddle.static.data(
        name="_matmul_y",
        shape=[1, hidden, hidden],
        dtype='float32')
    ele_y = paddle.static.data(
        name="_ele_y", shape=[hidden, ], dtype='float32')

    multi_fc_layer = MultiFCLayer(hidden, paddle.nn.ReLU)
    with paddle.static.amp.fp16_guard():
        out = multi_fc_layer(data, matmul_y, ele_y)
        loss = paddle.mean(out)
        paddle.static.append_backward(loss=loss)

build_strategy = paddle.static.BuildStrategy()
build_strategy.fuse_gemm_epilogue = True
program = paddle.static.CompiledProgram(smain_prog)
program = program.with_data_parallel(
    loss_name=loss.name,
    build_strategy=build_strategy,
    places=paddle.static.cuda_places())

# 3 subgraphs be fused to fused_gemm_epilogue_op
# 3 subgraphs be fused to fused_gemm_epilogue_grad
```

### CODE REVIEW

File: paddle/fluid/framework/ir/fuse_gemm_epilogue_pass.cc [hunk 1]
```diff
@@ -0,0 +1,443 @@
...
+
+  gpd(graph, handler);
+
+  AddStatis(found_ele_add_matmul_act_count);
+  return graph;
+}
+
+bool FuseGemmEpiloguePass::IsGemmFromLinear_(
+    const std::vector<int64_t> &x_shape, const std::vector<int64_t> &w_shape,
+    OpDesc *matmul_v2_op) const {
```
Comment (sneaxiy) [inline paddle/fluid/framework/ir/fuse_gemm_epilogue_pass.cc:420]: Why it is enough for GEMM from `paddle.nn.Linear`? For example, how about `[M, 1] * [N, K]`? Is it supported by cublasLt?

File: paddle/fluid/framework/ir/fuse_gemm_epilogue_pass.cc [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+    // is created by paddle.nn.Linear
+    auto matmul_grad_op_desc = matmul_grad_op->Op();
+    if (!IsGemmFromLinear_(matmul_grad_x_shape, matmul_grad_w_shape,
+                           matmul_grad_op_desc))
+      return;
+
+    OpDesc fused_gemm_epilogue_grad_op_desc(ele_add_grad_op->Op()->Block());
+    std::string activation_grad = "none";
+    fused_gemm_epilogue_grad_op_desc.SetType("fused_gemm_epilogue_grad");
+    fused_gemm_epilogue_grad_op_desc.SetInput("DOut",
```
Comment (sneaxiy) [inline paddle/fluid/framework/ir/fuse_gemm_epilogue_pass.cc:280]: How about `framework::GradVarName("Out")` instead of `DOut`? Same below.

File: paddle/fluid/framework/ir/fuse_gemm_epilogue_pass.cc [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+  };
+
+  gpd(graph, handler);
+
+  AddStatis(found_linear_act_count);
+  return graph;
+}
+
+ir::Graph *FuseGemmEpiloguePass::FuseLinearBwd(ir::Graph *graph,
+                                               bool is_first_gemm) const {
```
Comment (sneaxiy) [inline paddle/fluid/framework/ir/fuse_gemm_epilogue_pass.cc]: What is `is_first_gemm` for? What is the difference between the first and the others GEMM?

In the following code, I feel that `is_first_gemm == true` means the gradient of `X` is not needed?

File: paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+          true, false,
+          platform::errors::InvalidArgument(
+              "The activation attribute of fused_gemm_epilogue op should be"
+              " one of {\"none\", \"relu\", \"gelu\"}. But received %s."
+              "But received activation=%s.",
+              activation));
+    }
+
+    if (activation == "none" && auxiliary_key.size() > 0) {
+      VLOG(0)
```
Comment (sneaxiy) [inline paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc]: How about use `PADDLE_THROW` here? This may be a bug if the code is reachable?

File: paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+
+    if (ctx->HasOutput("DBias")) {
+      std::vector<int64_t> dbias_dims;
+      dbias_dims.push_back(y_dims[1]);
+      ctx->SetOutputDim("DBias", framework::make_ddim(dbias_dims));
+    }
+  }
+};
+
+class FusedGemmEpilogueGradOpMaker : public framework::OpProtoAndCheckerMaker {
```
Comment (sneaxiy) [inline paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc:305]: The op maker of the backward op is usually unnecessary.

File: paddle/fluid/operators/fused/fused_gemm_epilogue_op.h [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+namespace paddle {
+namespace operators {
+
+class EpilogueMetaData {
+ public:
+  memory::allocation::AllocationPtr auxiliary = nullptr;
+};
+
+class EpilogueSingleton {
```
Comment (sneaxiy) [inline paddle/fluid/operators/fused/fused_gemm_epilogue_op.h]: In my understanding, `EpilogueSingleton` is used to store a memory buffer which is written in the forward cublasLt API. This memory buffer must be passed to the backward cublasLt API without any modification. Therefore, you use a map to save the name-to-memory-buffer mapping here, and the name is the activation output name. Am I right?

I prefer to using something like `ReserveSpace` in batch_norm op. It is not encouraged to save the variable name inside the op attribute, which makes the graph dependency analysis, .etc difficult.

File: paddle/fluid/framework/ir/fuse_gemm_epilogue_pass.cc (hunk 3 above)
Comment (mingxu1067) [reply paddle/fluid/framework/ir/fuse_gemm_epilogue_pass.cc]: Done, change to `with_x_gradient`.

File: paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc (hunk 4 above)
Comment (mingxu1067) [reply paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc]: Done

File: paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc (hunk 5 above)
Comment (mingxu1067) [reply paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc:305]: Since `FusedGemmEpilogueGrad` is an independent op, not grad op of `FusedGemmEpilogue`, the maker is necessary.
`FusedGemmEpilogueGrad` includes scope of backward to `Linear(Act(X))`, but `FusedGemmEpilogue` includes `Act(Linear(X))`.

File: paddle/fluid/operators/fused/fused_gemm_epilogue_op.h (hunk 6 above)
Comment (mingxu1067) [reply paddle/fluid/operators/fused/fused_gemm_epilogue_op.h]: Done

File: paddle/fluid/framework/ir/fuse_gemm_epilogue_pass.cc (hunk 2 above)
Comment (mingxu1067) [reply paddle/fluid/framework/ir/fuse_gemm_epilogue_pass.cc:280]: It seems that forward ops are not allowed to have `@` in Input/Output names.
```
addle/fluid/eager/api/generated/fluid_generated/dygraph_forward_api.h:146:87: error: stray ‘@’ in program
EagerTensor fused_gemm_epilogue_grad_dygraph_function(const egr::EagerTensor& Out@GRAD,const egr::EagerTensor& X,const egr::EagerTensor& Y, const paddle::framework::AttributeMap& attr_map)
```

File: paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+    return framework::OpKernelType(data_type, ctx.GetPlace(), layout, library);
+  }
+};
+
+class FusedGemmEpilogueOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() override {
+    AddInput("X", "The input tensor X of Out = Act((X * Y) + bias).");
+    AddInput("Y", "The input tensor Y of Out = Act((X * Y) + bias).");
+    AddInput("bias", "The input tensor bias of Out = Act((X * Y) + bias).");
```
Comment (sneaxiy) [inline paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc]: `bias`->`Bias`?

File: paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+
+    AddOutput("out", "The output tensor Out of Out = Act((X * Y) + bias).");
```
Comment (sneaxiy) [inline paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc]: `out`->`Out`?

File: paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+    AddOutput("reserve_space",
```
Comment (sneaxiy) [inline paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc]: `reserve_space`->`ReserveSpace`?

File: paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+};
+
+class FusedGemmEpilogueGradOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() override {
+    AddInput("DOut",
+             "The input grad tensor to Out of Out = (Act(X) * Y) + bias");
+    AddInput("X", "The input tensor X of Out = (Act(X) * Y) + bias");
+    AddInput("Y", "The input tensor Y of Out = (Act(X) * Y) + bias");
+    AddInput("reserve_space",
```
Comment (sneaxiy) [inline paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc]: `reserve_space`->`ReserveSpace`?

File: paddle/fluid/framework/ir/fuse_gemm_epilogue_pass.cc [hunk 11]
```diff
@@ -0,0 +1,469 @@
...
+    // and mark it the act op is fused for backward epilogue fusion.
+    // That because cuBlasLt epilogue's restriction.
+    if (is_training) {
+      int divisor_of_n = activation == "relu" ? 128 : 8;
+      if (matmul_w_shape[1] % divisor_of_n) return;
+
+      VarDesc reserve_space(patterns::PDNodeName(scope_name, "reserve_space"));
+      auto *reserve_space_node = g->CreateVarNode(&reserve_space);
+
+      EpiloguePassActivationCache::Instance().InsertFusedActivation(
```
Comment (sneaxiy) [inline paddle/fluid/framework/ir/fuse_gemm_epilogue_pass.cc]: How about make `EpiloguePassActivationCache::Instance` to be a local variable instead of a singleton? I mean you can change the declaration of `FuseGemmEpiloguePass::FuseLinearActFwd` and `FuseGemmEpiloguePass::FuseLinearActBwd` to be:

```C++
ir::Graph *FuseGemmEpiloguePass::FuseLinearActFwd(
    ir::Graph *graph, const std::unordered_set<std::string> &act_types,
    bool is_training, bool is_act_grad_x_from_act, EpiloguePassActivationCache *cache) const;

ir::Graph *FuseGemmEpiloguePass::FuseLinearActBwd(
    ir::Graph *graph, const std::unordered_set<std::string> &act_grad_types,
    bool is_act_grad_x_from_act, const EpiloguePassActivationCache &cache) const;
```

File: paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc (hunk 8 above)
Comment (mingxu1067) [reply paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc]: Done

File: paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc (hunk 7 above)
Comment (mingxu1067) [reply paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc]: Done

File: paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc (hunk 9 above)
Comment (mingxu1067) [reply paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc]: Done

File: paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc (hunk 10 above)
Comment (mingxu1067) [reply paddle/fluid/operators/fused/fused_gemm_epilogue_op.cc]: Done

File: paddle/fluid/framework/ir/fuse_gemm_epilogue_pass.cc (hunk 11 above)
Comment (mingxu1067) [reply paddle/fluid/framework/ir/fuse_gemm_epilogue_pass.cc]: Done.
Made `EpiloguePassActivationCache`  as local variable and pass to `FuseLinearActFwd` and `FuseLinearActBwd`.
Using pointer rather than reference, due to request from pre-commit hooks.

**Review by sneaxiy — APPROVED** [review]
LGTM.

**Review by luotao1 — APPROVED** [review]
